app/abci.go

```diff
@@ -0,0 +1,160 @@
+package app
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lazyledger/lazyledger-app/x/lazyledgerapp/types"
+	abci "github.com/lazyledger/lazyledger-core/abci/types"
+	core "github.com/lazyledger/lazyledger-core/proto/tendermint/types"
+)
+
+// This file should contain all of the althered ABCI methods
+
+// PreprocessTxs fullfills the lazyledger-core version of the ACBI interface, by
+// performing basic validation for the incoming txs, and by cleanly separating
+// share messages from transactions
+// todo(evan): refactor out a for loop.
+func (app *App) PreprocessTxs(txs abci.RequestPreprocessTxs) abci.ResponsePreprocessTxs {
+	squareSize := app.SquareSize()
+	shareCounter := uint64(0)
+	var shareMsgs []*core.Message
+	var processedTxs [][]byte
+	for _, rawTx := range txs.Txs {
+		// decode the Tx
+		tx, err := app.txConfig.TxDecoder()(rawTx)
+		if err != nil {
+			continue
+		}
+
+		// don't do anything if the transaction doesn't contain a PayForMessage sdk.Msg
+		if !hasWirePayForMessage(tx) {
+			processedTxs = append(processedTxs, rawTx)
+			continue
+		}
+
+		// only support transactions that contain a single sdk.Msg for now
+		if len(tx.GetMsgs()) != 1 {
+			continue
+		}
+
+		msg := tx.GetMsgs()[0]
+
+		// run basic validation on the transaction
+		err = tx.ValidateBasic()
+		if err != nil {
+			continue
+		}
+
+		// process and validate the message
+		coreMsg, signedTx, err := app.processMsg(msg)
+		if err != nil {
+			continue
+		}
+
+		// execute the tx in runTxModeDeliver mode (3)
+		// execution includes all validation checks burning fees
+		// currently, no fees are burned
+		_, _, err = app.BaseApp.TxRunner()(3, rawTx)
+		if err != nil {
+
+			continue
+		}
+
+		// increment the share counter by the number of shares taken by the message
+		sharesTaken := uint64(len(coreMsg.Data) / 256)
+		shareCounter += sharesTaken
+
+		// if there are too many shares stop processing and return the transactions
+		// this is related to issues 67 and 77 of lazyledger-core
+		if shareCounter > squareSize {
+			break
+		}
```

Comment:
Regarding squareSize: for a `k` times `k` matrix, does it represent `k` or `k*k`?

Reply:
it's currently only k, but should definitely be k*k